jar/jar.go

```diff
@@ -334,31 +336,114 @@ var (
 	//
 	// Since this is so brittle, we're keeping the above rule that can reliably and
 	// non-brittle-ey detect <2.15 as a back up.
-	log4j216Detector = []byte("isJndiEnabled")
+	log4J216Pattern = newBytePattern([]bytePatternSegment{{literal: []byte("isJndiEnabled")}})
 )
 
-func matchesLog4JYARARule(b []byte) bool {
-	start := 0
+func (c *checker) checkClass(filename string, r io.Reader, buf []byte) error {
+	if !c.hasLookupClass && strings.Contains(filename, "JndiLookup.class") {
+		c.hasLookupClass = true
+	}
+	checkForOldJndiManagerConstructor := !c.hasOldJndiManagerConstructor && strings.Contains(filename, "JndiManager")
+	checkJndiManagerVersion := strings.Contains(filename, "JndiManager.class")
+	if !checkForOldJndiManagerConstructor && !checkJndiManagerVersion {
+		return nil
+	}
+	if checkJndiManagerVersion {
+		c.seenJndiManagerClass = true
+	}
+	yaraMatcher := newByteMatcher(log4JYARAPattern)
+	twoSixteenMatcher := newByteMatcher(log4J216Pattern)
 	for {
-		i := bytes.Index(b[start:], log4JYARAPrefix)
-		if i < 0 {
-			return false
+		n, err := r.Read(buf)
+		for i := 0; i < n; i++ {
+			if checkForOldJndiManagerConstructor && !c.hasOldJndiManagerConstructor {
+				if yaraMatcher.next(buf[i]) {
+					c.hasOldJndiManagerConstructor = true
+				}
+			}
+			if checkJndiManagerVersion && !c.isAtLeastTwoDotSixteen {
+				if twoSixteenMatcher.next(buf[i]) {
+					c.isAtLeastTwoDotSixteen = true
+				}
+			}
 		}
-		n := i + len(log4JYARAPrefix)
-		if len(b) <= n {
-			return false
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("reading file %s: %w", filename, err)
 		}
-		j := bytes.Index(b[n:], log4JYARASuffix)
-		if j < 0 {
+	}
+	return nil
+}
+
+// byteMatcher is a state machine that reports matches for byte patterns.
+//
+// Must not be used once next returns true.
+type byteMatcher struct {
+	pattern   *bytePattern
+	byteIndex int
+	segmIndex int
+	found     bool
+}
+
+func newByteMatcher(pattern *bytePattern) *byteMatcher {
+	return &byteMatcher{pattern: pattern}
+}
+
+// next moves the state machine to the next state.
+//
+// Returns true if a match was found. Once a match is
+// found, it will only return true after that point.
+func (m *byteMatcher) next(b byte) bool {
```

Comment:
It's not clear how this handles mismatches where you have to rewind the stream. E.g. consider the pattern

```
[]byte{0x01, 0x01, 0x02}
```

Then a byte stream

```
[]byte{0x01, 0x01, 0x01, 0x02}
```

Wouldn't you match the first two elements, reset when 0x01 != 0x02, then not catch the match?

Reply:
that's a good point about byte patterns in general, but in this specific case, because there are no consecutively repeating bytes in the patterns we're looking for, I don't believe it's possible to have these false starts. put another way, I think that https://pkg.go.dev/rsc.io/binaryregexp#Regexp.LiteralPrefix would always return the complete pattern.

the matcher I wrote is very, very narrow; perhaps it shouldn't even be labeled as a matcher at all, or that it should check on initialization that there are no repeats in the pattern prefix or suffix.

alternatively I can make the matcher more powerful and general by backing it with a mini DFA. I'd use rsc.io/binaryregexp in a heartbeat, but I don't think it's possible to check both of these conditions in a single pass using that API.

WDYT?

Reply:
ah, but there is actually a bug in here anyway. yes, definitely needs tests.

Reply:
oops, no. there definitely is a way to do this with binaryregexp. but I don't know if it's better. I'm going to try that and try to avoid this question entirely.

Reply:
Sounds good :)

Please re-run the benchmark and post the results. If it's not a huge gain, it might be worth sticking with our naive io.ReadAll strategy?

Reply:
now it uses binaryregexp